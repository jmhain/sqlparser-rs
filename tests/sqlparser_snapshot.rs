// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#![warn(clippy::all)]

use sqlparser::ast::Statement;
use sqlparser::dialect::{Dialect, GenericDialect};
use sqlparser::parser::Parser;
use std::path::Path;

fn test_dialect(path: &Path, dialect: &dyn Dialect) -> datatest_stable::Result<()> {
    let sql = std::fs::read_to_string(path)?;

    let stmts = Parser::parse_sql(dialect, &sql)?;
    let name = path
        .strip_prefix(Path::new("tests/queries"))?
        .to_str()
        .unwrap();
    insta::assert_yaml_snapshot!(name, stmts);

    // Rendering the AST to a string and re-parsing it should yield the same AST.
    for stmt in stmts {
        let [round_trip_stmt]: [Statement; 1] = Parser::parse_sql(dialect, &stmt.to_string())?
            .try_into()
            .map_err(|_| "expected a single statement")?;
        pretty_assertions::assert_eq!(stmt, round_trip_stmt);
    }

    Ok(())
}

fn test_generic(path: &Path) -> datatest_stable::Result<()> {
    test_dialect(path, &GenericDialect)
}

datatest_stable::harness!(test_generic, "tests/queries/generic", r"^.*.sql",);
