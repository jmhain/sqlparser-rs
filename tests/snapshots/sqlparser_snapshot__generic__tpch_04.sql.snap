---
source: tests/sqlparser_snapshot.rs
expression: stmts
---
- Query:
    with: ~
    body:
      Select:
        distinct: ~
        top: ~
        projection:
          - UnnamedExpr:
              Identifier:
                value: o_orderpriority
                quote_style: ~
          - ExprWithAlias:
              expr:
                Function:
                  name:
                    - value: count
                      quote_style: ~
                  args:
                    - Unnamed: Wildcard
                  filter: ~
                  null_treatment: ~
                  over: ~
                  distinct: false
                  special: false
                  order_by: []
              alias:
                value: order_count
                quote_style: ~
        into: ~
        from:
          - relation:
              Table:
                name:
                  - value: orders
                    quote_style: ~
                alias: ~
                args: ~
                with_hints: []
                version: ~
                partitions: []
            joins: []
        lateral_views: []
        selection:
          BinaryOp:
            left:
              BinaryOp:
                left:
                  BinaryOp:
                    left:
                      Identifier:
                        value: o_orderdate
                        quote_style: ~
                    op: GtEq
                    right:
                      TypedString:
                        data_type: Date
                        value: 1993-07-01
                op: And
                right:
                  BinaryOp:
                    left:
                      Identifier:
                        value: o_orderdate
                        quote_style: ~
                    op: Lt
                    right:
                      BinaryOp:
                        left:
                          TypedString:
                            data_type: Date
                            value: 1993-07-01
                        op: Plus
                        right:
                          Interval:
                            value:
                              Value:
                                SingleQuotedString: "3"
                            leading_field: Month
                            leading_precision: ~
                            last_field: ~
                            fractional_seconds_precision: ~
            op: And
            right:
              Exists:
                subquery:
                  with: ~
                  body:
                    Select:
                      distinct: ~
                      top: ~
                      projection:
                        - Wildcard:
                            opt_exclude: ~
                            opt_except: ~
                            opt_rename: ~
                            opt_replace: ~
                      into: ~
                      from:
                        - relation:
                            Table:
                              name:
                                - value: lineitem
                                  quote_style: ~
                              alias: ~
                              args: ~
                              with_hints: []
                              version: ~
                              partitions: []
                          joins: []
                      lateral_views: []
                      selection:
                        BinaryOp:
                          left:
                            BinaryOp:
                              left:
                                Identifier:
                                  value: l_orderkey
                                  quote_style: ~
                              op: Eq
                              right:
                                Identifier:
                                  value: o_orderkey
                                  quote_style: ~
                          op: And
                          right:
                            BinaryOp:
                              left:
                                Identifier:
                                  value: l_commitdate
                                  quote_style: ~
                              op: Lt
                              right:
                                Identifier:
                                  value: l_receiptdate
                                  quote_style: ~
                      group_by:
                        Expressions: []
                      cluster_by: []
                      distribute_by: []
                      sort_by: []
                      having: ~
                      named_window: []
                      qualify: ~
                  order_by: []
                  limit: ~
                  limit_by: []
                  offset: ~
                  fetch: ~
                  locks: []
                negated: false
        group_by:
          Expressions:
            - Identifier:
                value: o_orderpriority
                quote_style: ~
        cluster_by: []
        distribute_by: []
        sort_by: []
        having: ~
        named_window: []
        qualify: ~
    order_by:
      - expr:
          Identifier:
            value: o_orderpriority
            quote_style: ~
        asc: ~
        nulls_first: ~
    limit: ~
    limit_by: []
    offset: ~
    fetch: ~
    locks: []

